### PR TITLE
8334215: serviceability/dcmd/thread/PrintMountedVirtualThread.java failing with JTREG_TEST_THREAD_FACTORY=Virtual

### PR DESCRIPTION
**Pull request by txominpelu**
Follow up to https://github.com/openjdk/jdk/pull/19482 that was causing issues when the PrintMountedVirtualTest.java was
running with `JTREG_TEST_THREAD_FACTORY=Virtual` in the loom repo. 

- Fixes issues where the test observes the thread during transitions.
- Fixes a potential issue in the test where CountDownLatch.countDown unparks the main (virtual) thread and the main thread observes the dummy thread is transition .

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334215](https://bugs.openjdk.org/browse/JDK-8334215): serviceability/dcmd/thread/PrintMountedVirtualThread.java failing with JTREG_TEST_THREAD_FACTORY=Virtual (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19744/head:pull/19744` \
`$ git checkout pull/19744`

Update a local copy of the PR: \
`$ git checkout pull/19744` \
`$ git pull https://git.openjdk.org/jdk.git pull/19744/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19744`

View PR using the GUI difftool: \
`$ git pr show -t 19744`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19744.diff">https://git.openjdk.org/jdk/pull/19744.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19744#issuecomment-2172839937)

### CODE REVIEW

File: src/hotspot/share/runtime/threads.cpp [hunk 1]
```diff
@@ -1331,9 +1331,11 @@ void Threads::print_on(outputStream* st, bool print_stacks,
         if (thread_oop != nullptr) {
           if (p->is_vthread_mounted()) {
             const oop vt = p->vthread();
-            assert(vt != nullptr, "vthread should not be null when vthread is mounted");
-            st->print_cr("   Mounted virtual thread \"%s\" #" INT64_FORMAT, JavaThread::name_for(vt), (int64_t)java_lang_Thread::thread_id(vt));
-            p->print_vthread_stack_on(st);
+            if (vt != thread_oop) {
+              assert(vt != nullptr, "vthread should not be null when vthread is mounted");
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/threads.cpp:1335]: I think the assert still belongs in the original position doesn't it? Or could the problem we hit here cause a transient null to appear as well?

File: src/hotspot/share/runtime/threads.cpp [hunk 2]
```diff
@@ -1331,9 +1331,11 @@ void Threads::print_on(outputStream* st, bool print_stacks,
         if (thread_oop != nullptr) {
           if (p->is_vthread_mounted()) {
             const oop vt = p->vthread();
-            assert(vt != nullptr, "vthread should not be null when vthread is mounted");
-            st->print_cr("   Mounted virtual thread \"%s\" #" INT64_FORMAT, JavaThread::name_for(vt), (int64_t)java_lang_Thread::thread_id(vt));
-            p->print_vthread_stack_on(st);
+            if (vt != thread_oop) {
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/threads.cpp:1334]: We need a comment explaining why we have to check for this as without knowing the implementation quirks it seems non-sensical for a thread to have mounted itself!

File: test/hotspot/jtreg/serviceability/dcmd/thread/PrintMountedVirtualThread.java [hunk 3]
```diff
@@ -41,11 +40,13 @@ public class PrintMountedVirtualThread {
 
     public void run(CommandExecutor executor) throws InterruptedException {
         var shouldFinish = new AtomicBoolean(false);
-        var started = new CountDownLatch(1);
+        var started = new AtomicBoolean();
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/dcmd/thread/PrintMountedVirtualThread.java:43]: Not sure how this change make things better? Why would we want to sleep and poll rather than park until signalled?

File: src/hotspot/share/runtime/threads.cpp (hunk 2 above)
Comment (AlanBateman) [reply src/hotspot/share/runtime/threads.cpp:1334]: > We need a comment explaining why we have to check for this as without knowing the implementation quirks it seems non-sensical for a thread to have mounted itself!

Just to note that JavaThread._vthread is always the "current thread". It's the virtual thread when mounted, the carrier (as in itself) when there is no virtual thread mounted. It's one this way so that the C2 instrinic for currentThread doesn't need a branch/test.

File: src/hotspot/share/runtime/threads.cpp (hunk 1 above)
Comment (txominpelu) [reply src/hotspot/share/runtime/threads.cpp:1335]: I think that, even if based on [this](https://github.com/openjdk/jdk/pull/19482#issuecomment-2166116062) comment from @AlanBateman's in some cases the virtual thread can be mounted but `_vthread` can refer to the carrier thread, what you're saying is correct and in both cases `_vthread` should still not be a null pointer:

- Either it's equal to `thread_oop` and we've already checked that it's not `nullptr`.
- Or it's different to `thread_oop` and it refers to an actual virtual thread that should not be null.

File: src/hotspot/share/runtime/threads.cpp (hunk 2 above)
Comment (txominpelu) [reply src/hotspot/share/runtime/threads.cpp:1334]: I wasn't aware of that, and I was going to suggest to add that information as documentation, but it seems to be already documented here: https://github.com/openjdk/jdk/blob/ad0b54d429fdbd806c09aa06bb42f1ed4a0297e8/src/hotspot/share/runtime/javaThread.hpp#L95

Comment (txominpelu) [reply src/hotspot/share/runtime/threads.cpp:1335]: I've still clarified why the if is required in 9a954efe164d9caee31494773c0eecc8ff69223f

File: src/hotspot/share/runtime/threads.cpp (hunk 1 above)
Comment (txominpelu) [reply src/hotspot/share/runtime/threads.cpp]: Moving the assert before the `if` in: 9a954efe164d9caee31494773c0eecc8ff69223f

File: test/hotspot/jtreg/serviceability/dcmd/thread/PrintMountedVirtualThread.java (hunk 3 above)
Comment (txominpelu) [reply test/hotspot/jtreg/serviceability/dcmd/thread/PrintMountedVirtualThread.java:43]: @AlanBateman explained to me why that is possible in https://github.com/openjdk/jdk/pull/19482#issuecomment-2166116062 and https://github.com/openjdk/jdk/pull/19482#issuecomment-2167374446. 

Alan can correct me if I'm wrong, but the dummy thread trying to unpark the main (virtual) thread needs to run as the carrier to avoid nested parking. See https://github.com/openjdk/jdk/blob/412e306d81209c05f55aee7663f7abb80286e361/src/java.base/share/classes/java/lang/VirtualThread.java#L707

File: src/hotspot/share/runtime/threads.cpp [hunk 4]
```diff
@@ -1332,8 +1332,11 @@ void Threads::print_on(outputStream* st, bool print_stacks,
           if (p->is_vthread_mounted()) {
             const oop vt = p->vthread();
             assert(vt != nullptr, "vthread should not be null when vthread is mounted");
-            st->print_cr("   Mounted virtual thread \"%s\" #" INT64_FORMAT, JavaThread::name_for(vt), (int64_t)java_lang_Thread::thread_id(vt));
-            p->print_vthread_stack_on(st);
+            if (vt != thread_oop) {
+              // JavaThread._vthread can refer to the carrier thread. Print only if _vthread refers to a virtual thread.
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/threads.cpp:1336]: Nit: please put the comment before the `if`

**Comment by txominpelu** [comment]
Thanks for your comments @dholmes-ora and @txominpelu . This PR is ready for another round of reviews.

File: test/hotspot/jtreg/serviceability/dcmd/thread/PrintMountedVirtualThread.java (hunk 3 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/serviceability/dcmd/thread/PrintMountedVirtualThread.java:43]: So IIUC you've changed the test to try to avoid the problem that was causing the test to fail, as well as changing the code so the test would not fail. But now we won't/may-not actually test that. ?? I guess I want to see that the original failing test now passes with the fix.

**Comment by AlanBateman** [comment]
I looked at the changes in the latest revision (9a954efe) and they look okay, it basically skips printing the mounted virtual thread when that virtual thread is interacting with the scheduler or timer mechanism, the so called "temporary transitions" that are needed to ensure the correct identity when submitting tasks to the scheduler, and to avoid deadlock when submitting tasks to the scheduler requires locking. It also changes the test so testing with JTREG_TEST_THREAD_FACTORY=Virtual won't observe the dummy thread in a temporary transition.

We can improve on this in the loom repo as JavaThread has the thread ID of the mounted virtual thread. Note that the pre-existing "Carrying ..." line has the same issue as we've been discussing here, it just hasn't been noticed perhaps.

One thing that I think we should do is remove the thread name from the "Mounted ..." line. I initially thought this was a good idea but there is no reference to this when a virtual thread is mounted and interacting with the scheduler. Virtual threads don't have a name by default so it's not a big a loss as you might think. We can of course re-visit this in the future.

File: test/hotspot/jtreg/serviceability/dcmd/thread/PrintMountedVirtualThread.java (hunk 3 above)
Comment (txominpelu) [reply test/hotspot/jtreg/serviceability/dcmd/thread/PrintMountedVirtualThread.java:43]: The fix in the threads.cpp code prevents the `Mounted Thread` statement to be printed when the command is incapable of actually printing the Vthread stack because the Vthread is running in the context of the carrier. A vthread running in the context of the carrier can happen when the vthread is undergoing a transition, for example when the vthread  unparks another virtual thread.

Even with that fix, the original test was invalid because it expected to always see the mounted thread in the ouptut, but waking up the main virtual thread after `started.await()` required that the dummy vthread had to unpark the main vthread, what requires a that dummy unparks main and has to transition to carrier. After the fixed done in `threads.cpp` this would result in the mounted thread not showing up in the output in some cases.

To sum up, we needed both changes but the `threads.cpp` change did not make the original test pass.

File: src/hotspot/share/runtime/threads.cpp (hunk 4 above)
Comment (txominpelu) [reply src/hotspot/share/runtime/threads.cpp]: Thanks ! I've fixed that in: 05c0f75acd6a37815146c7a7e1b30aede47ab77a

**Comment by txominpelu** [comment]
> I looked at the changes in the latest revision ([9a954ef](https://github.com/openjdk/jdk/commit/9a954efe164d9caee31494773c0eecc8ff69223f)) and they look okay, it basically skips printing the mounted virtual thread when that virtual thread is interacting with the scheduler or timer mechanism, the so called "temporary transitions" that are needed to ensure the correct identity when submitting tasks to the scheduler, and to avoid deadlock when submitting tasks to the scheduler requires locking. It also changes the test so testing with JTREG_TEST_THREAD_FACTORY=Virtual won't observe the dummy thread in a temporary transition.
> 
> We can improve on this in the loom repo as JavaThread has the thread ID of the mounted virtual thread. Note that the pre-existing "Carrying ..." line has the same issue as we've been discussing here, it just hasn't been noticed perhaps.
> 
> One thing that I think we should do is remove the thread name from the "Mounted ..." line. I initially thought this was a good idea but there is no reference to this when a virtual thread is mounted and interacting with the scheduler. Virtual threads don't have a name by default so it's not a big a loss as you might think. We can of course re-visit this in the future.

Thanks ! I've removed the virtual thread name from the "Mounted .." line in: af85b2617c97eeb31ce347f42b019c261bc141f1

**Comment by txominpelu** [comment]
Thanks again ! @AlanBateman @dholmes-ora, I've responded to your latest comments and incorporated your latest suggestions.

**Review by dholmes-ora — APPROVED** [review]
Okay. I still have some general concerns about the underlying implementation issues, but as long as the test will now pass reliably that is okay.

FWIW I would have kept the vt name just in case it does have one. But as Alan said that can be added in later if desired.

Thanks.

**Comment by AlanBateman** [comment]
> FWIW I would have kept the vt name just in case it does have one. But as Alan said that can be added in later if desired.

I suggested we drop this for now because the thread dump can happen at times when there isn't a reference to the virtual Thread, so can't get the thread name, at least not easily. Definitely something to come back to.  The changes in the loom repo means we at least have the thread ID of the mounted virtual thread at all times so this means we can fix the "Carrying .." line and also adjust the changes in this PR to remove the check for the self reference, meaning the "Mounted ..." line and the stack trace won't skipped when the thread dump is taken during a transition.

**Comment by dholmes-ora** [comment]
> the thread dump can happen at times when there isn't a reference to the virtual Thread

I don't follow. We have checked vt is not null and not the carrier, so we do have it.

**Comment by AlanBateman** [comment]
> I don't follow. We have checked vt is not null and not the carrier, so we do have it.

My comment was about the case where the thread dump happens during a transition so sees the self-reference. In the loom repo the JavaThread has the thread ID of the mounted virtual thread so it can support this case, it doesn't need this check. So improvements in the future when the changes get the main line.
